### PR TITLE
Filter out Microsoft.FSharp.Core from Add Open Codefix

### LINKED ISSUE #2511: Wrong open namespace suggestion in case of FS0043
image

@cartermp it looks like it's been introduced by your last PR.

### PR DESCRIPTION
**Pull request by cartermp**
This fixes #2511.  I'm not sure if this is the best overall approach, but given that `tryFindInsertionContext` is only ever called from the Add Open Codefix it's safe.  I can't think of any reason why we'd *want* `Microsoft.FSharp.Core` to show up in that list since it's always referenced by default.

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
@cartermp have you found why it popped up? An Entity must trigger it, I think it was operator +. So I think it 
* sees the 0043 error
* sees `op_Plus` (or whatever compilated name + has) 
* decides that it's not in scope
* searches all referenced assemblies (and FSharpCore of course)
* finds op_Plus in those two namespaces / modules 
* suggests to open them

So it's not a FSharp.Core specific problem, it's a result of triggering the code fix on 0043 diagnostic id.

**Comment by cartermp** [comment]
Right, but not triggering on that error code will end up in the same situation where we were before, where user-defined operators wouldn't trigger the codefix (e.g., `>=>` from `Suave.Operators`).

In this case, we were getting a bad result because FS0043 is *also* used when the types mismatch on the operator (e.g., `1 + 2.0`).  I decided that filtering out Microsoft.FSharp.Core from entities to be created made sense for all cases though, since there isn't any circumstance where we'd want to pull that in for a codefix.  Maybe the better approach is to filer Microsoft.FSharp.Core from `AssemblyContentProvider.GetAllEntitiesInProjectAndReferencedAssemblies`, since that's also only ever used in the Add Open Codefix.

**Comment by vasily-kirichenko** [comment]
@cartermp we could get all opened namespaces and modules at the cursor point and use them to filter out wrong suggestions.

**Comment by vasily-kirichenko** [comment]
@cartermp See my PR https://github.com/Microsoft/visualfsharp/pull/2515, I think it's a better approach.

**Comment by KevinRansom** [comment]
@vasily-kirichenko @cartermp 

Okay ... 

I spent some time chatting with Phillip about this in the office today ...

image
This is pretty much the simplest interesting repro I could come up with:

````
module ANamespace =
    let (*) (_:double) (_:float) = 37.2

module ANotherNamespace =
    let (*) (_:int) (_:float) = 37.2

let y = 3 * 5.0
````
Produces this:

temp

The best suggestion is AnotherNamespace which is listed 2nd ... this is the best suggestion because it has the operator (*) and the types match ... So it fixes the problem.

The second best suggestion is ANamespace because it's operator star has at least one valid type argument and so if the developer selects ANamespace instead he will see this:

temp

FSharp.Core and FSharp.Core.checked are okay suggestions, because they contain an operator (*) however both arguments are incorrect and now intellisense will show the same errors as before, requiring the developer to get the type arguments correct.

There is also another possible fix though and that is to implement the operator ...

So to summarise Filtering out fsharp.core is not correct ... but the current implementation needs to ensure that an exact match by types appears  at the top of the list,

I hope this helps

Kevin

**Comment by cartermp** [comment]
Just to clarify, Microsoft.FSharp.Core is *not* an okay suggestion, since it's already open by default.  Anything sitting behind a module inside of FSharp.Core is a candidate for the codefix, but Microsoft.FSharp.Core should never be one.  That should be addressed in #2515, though.

**Comment by cartermp** [comment]
Closing this in favor of #2515 though, since it's a smarter approach than singling out Microsoft.FSharp.Core, IMO.
